### ISSUE TITLE
Selenium on Edge doesn't invoke onchange() when calling SelectElement.SelectByValue()

### ISSUE THREAD

**Issue by Omorocker** [SeleniumHQ/selenium#1621]
I have the following HTML:
```
<select onchange="redirect(this)" style="position:relative;top:3px">        
    <option value="?locale=fr">Français</option>        
    <option value="?locale=en">English</option>     
    <option value="?locale=zh">Chinese</option>     
</select>
```
In my Selenium C# test case, I do:
```
[FindsBy(How = How.XPath, Using = "//select[not(option[1][@value='-1'])]")]
protected IWebElement ComboBoxLanguageInternal { get; set; }
protected SelectElement ComboBoxLanguage => new SelectElement(ComboBoxLanguageInternal);

this.SelectLanguage("en");

public void SelectLanguage(string locale)
{
    ComboBoxLanguage.SelectByValue("?locale=" + locale);
}
```
This works for all browsers and platforms. The language select element is switched from "Français" to "English" and the JavaScript function redirect() is executed. The only exception is MS Edge on Win10 Desktop: Here, the value is also switched to "English" but the JavaScript onchange event is not invoked.

**Comment by InstyleVII** [comment]
Hey, I'm the PM for WebDriver on the Edge team. Just wanted to let you know that this is a known bug we have on us to fix. If your in the Insiders program and getting builds you'll see a fix there first.

There are a few bugs around change events not firing that are known, specifically:
Clear Command not firing it
Changing select in a multi-select doesn't not firing it

**Comment by lukeis** [comment]
@InstyleVII thanks for the update :)

@Omorocker Edge issues are tracked by Microsoft here: https://connect.microsoft.com/  I would normally say log a bug there instead, except it seems the bug is already fixed in their beta (insiders) build.

**Comment by InstyleVII** [comment]
Just to clarify this isn't fixed yet, we're working on it. We will however have a pretty substantial update out to Insiders soon. This won't be fixed in it but a lot of general stability fixes will be.

**Comment by titusfortner** [comment]
That's an update to the Microsoft WebDriver Fall 2015 Update? Is there an ETA? I've been holding off on doing more Edge testing until that comes out. :)

**Comment by InstyleVII** [comment]
No, to the WebDriver for Windows Insiders download located here:
https://www.microsoft.com/en-us/download/details.aspx?id=48740

I've been holding off on updating the Insiders drop as all of the Insider builds we have flighted so far haven't had any updates to WebDriver. The next flight *should* have updates depending on timing. If it does I'll update the Insiders WebDriver server as well. It'll mainly be stability fixes for issues regarding crashing and race conditions. No new big features.

**Comment by Omorocker** [comment]
Hey Clayton,

thanks for getting in touch.

And while we’re at it, could you please have a quick look at https://github.com/SeleniumHQ/selenium/issues/1622 ? This has been closed by the Selenium people, as they say it‘s an Edge bug and should be posted on http://connect.microsoft.com/ . Would you agree? Or is this a know bug, too?

BTW, your WebDriver 2.0 is a major improvement to previous versions. It’s already very close to professionally usable. So keep up the good work!

Bye…
Kim

Von: Clayton Martin [mailto:notifications@github.com]
Gesendet: Donnerstag, 11. Februar 2016 01:11
An: SeleniumHQ/selenium
Cc: Homann, Kim
Betreff: Re: [selenium] Selenium on Edge doesn't invoke onchange() when calling SelectElement.SelectByValue() (#1621)


Hey, I'm the PM for WebDriver on the Edge team. Just wanted to let you know that this is a known bug we have on us to fix. If your in the Insiders program and getting builds you'll see a fix there first.

There are a few bugs around change events not firing that are known, specifically:
Clear Command not firing it
Changing select in a multi-select doesn't not firing it

**Comment by InstyleVII** [comment]
Hey @Omorocker,

Wow good test! Found a few bugs, some known, some unknown when running your test.

1. Timing issue - Bug filed - Sometimes we fail to process this command:
`driver.FindElement(By.CssSelector("div.button_warenkorb_im_kontext>a")).Click();`
2. Markup issue - Bug filed - When the markup has whitespace we fail to find the element by LinkText. Running your test on markup like this worked:
`<a href=""><button><span><span>zur Kasse</span></span></button></a>`
But add in some formatting like returns after each element and it fails
3. Text value contains whitespace - Bug known - Re-prioritized a bug in our backlog for whitespace in the Text property of an element. Other drivers remove the whitespace while we don't, so in Edge you get "\r\n\r\nzur Kasse\r\n " vs. Chrome and Firefox which return "zur Kasse"

I've filed bugs for the first two and prioritized the third higher. I'd keep an eye on our Insider preview for updates to WebDriver. I mentioned it in the above conversation but that'll have the latest bug fixes for WebDriver.

Thanks,
Clay

**Comment by Omorocker** [comment]
Hi Clayton,

thank you!


1.)    This is a scrolling issue, which I found a workaround for: I altered the Click() method especially for Edge to try Click(), catch any exception, and in the exception handler I call MoveToElement() to scroll to the element. But that is not enough. I also need to scroll down a few pixels more before the element gets really clickable. I call JavaScript „window.scrollBy(0,10);“ to achieve this. Then I call Click() again, and I’m fine.


2.)    I’m using XPath as a workaround. As I’m using PageFactory, I had to do it for all browsers, which is not good. But I’m fine for the moment.


3.)    I added some custom handling for Edge as a workaround here, too. Instead of Selenium „Text“ property, I’m using a „Text()“ extension method:
       public static string Text(this IWebElement element, TestParams testParams)
       {
           if (testParams.Target.IsEdge)
           {
               var strings = element.Text.Trim().Split(new []{"\r\n" }, StringSplitOptions.RemoveEmptyEntries);
               for (int cnt = 0; cnt < strings.Length; cnt++)
               {
                   strings[cnt] = strings[cnt].Trim();
               }

               return string.Join("\r\n", strings);
           }

           return element.Text;
       }

Final thing left before all my test cases hopefully go green with Edge is SendKeys() with the „@“ character, which doesn’t work with Edge. Thus you cannot enter any email addresses in any text fields. But I’m going to use JavaScript as a workaround as propsed in http://stackoverflow.com/questions/35342869/ .

Bye…
Kim
